### PR TITLE
Support :title_view instances defining their own height for section headers

### PR DESCRIPTION
**Pull request by markrickert**
## :tada:  :tada: It's like magic! :tada: :tada: 

 No more manually specifying `:title_view_height`. Just define a `height` method in your custom title view class, change the height after `title=` is called to be whatever it needs to be and then return `self.frame.size.height` in your `height` method!

Here's an example implementation (with some RMQ mixed in since you should be using [redpotion](https://github.com/infinitered/redpotion) anyways :smile:)

```ruby
class TableSectionHeaderView < UIView
  # This will create table section header view with a UILabel
  # with the correct font (defined in application_stylesheet.rb)
  # for section headers.
  #
  # Instructions on how to use:
  #
  # def table_data
  #   [{
  #     title_view: TableSectionHeaderView,
  #     title: "Whatever string you want"
  #   }]
  # end

  def title=(t)
    # Apply a style to the view instance
    @v = find(self).apply_style(:table_section_header_view)

    # Create the UILabel and set the data to the text provided
    title = append(UILabel, :table_section_header_label)
    title.data = t

    # Resize the title frame to fit the text
    title.style{|st| st.resize_height_to_fit }

    # Resize the view instance height
    @v.style do |st|
      st.frame = {
        h: title.frame.size.height + 15
      }
    end
  end

  # Return the height of the view
  def height
    @v.frame.size.height
  end
end
```

### CODE REVIEW

**Comment by markrickert** [comment]
All tests passing... CircleCI just isn't configured.

**Comment by markrickert** [comment]
Fixing merge issue right now...

**Comment by markrickert** [comment]
Fixed.

```
311 specifications (640 requirements), 0 failures, 0 errors
```

File: lib/ProMotion/table/table.rb [hunk 1]
```diff
@@ -299,7 +299,13 @@ def tableView(_, viewForHeaderInSection: index)
     def tableView(_, heightForHeaderInSection: index)
       section = promotion_table_data.section(index)
       if section[:title_view] || section[:title].to_s.length > 0
-        section[:title_view_height] || tableView.sectionHeaderHeight
+        if section[:title_view_height]
+          section[:title_view_height]
+        elsif tableView(_, viewForHeaderInSection: index).respond_to?(:height)
+          tableView(_, viewForHeaderInSection: index).height
```
Comment (markrickert) [inline lib/ProMotion/table/table.rb]: @jamonholmgren is there a way we can optimize this so we're not calling the method twice? maybe something like:

```ruby
elsif header_class = tableView(_, viewForHeaderInSection: index) && header_class.respond_to?(:height)
  header_class.height
end
```

I've never been good at optimizing stuff like this.

Comment (markrickert) [inline lib/ProMotion/table/table.rb]: Think i got it - 

```ruby
elsif (section_header = tableView(_, viewForHeaderInSection: index)) && section_header.respond_to?(:height)
  section_header.height
end
```

File: app/test_views/dynamic_height_title_view.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+class DynamicHeightTitleView < UIView
+  attr_accessor :title
+
+  def title=(t)
```
Comment (jamonholmgren) [inline app/test_views/dynamic_height_title_view.rb]: Can we change this to the new `on_load` hook?

Comment (markrickert) [inline app/test_views/dynamic_height_title_view.rb]: I need some way to actually set the title FROM the test and there's no way to do that unless I repurpose the `title=` call. That method gets called if there's a title string specified as well as the `:title_view` custom class.

**Comment by jamonholmgren** [comment]
I'll update it to use on_load

**Comment by jamonholmgren** [comment]
I can't wait to use this!

**Comment by markrickert** [comment]
Makes it a lot easier to do custom fonts in section header views now :)

**Pull request merged**